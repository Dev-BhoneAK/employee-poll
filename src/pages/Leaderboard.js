import {connect} from "react-redux";

const Leaderboard = ({users, userIds, authedUser}) => {

    return (
        <div className={"table-container"}>
            <table>
                <thead>
                    <tr>
                        <th>Users</th>
                        <th>Created</th>
                        <th>Answered</th>
                    </tr>
                </thead>
                <tbody>
                    {userIds.map(userId => (
                        <tr key={userId} style={authedUser === userId ? {color: "#5885DD"}: {}}>
                            <td>
                                <div className="user-data">
                                    <img src={users[userId].avatarURL} alt="Avatar" className={"leaderboard-avatar"} />
                                    <div >
                                        <h4><b>{users[userId].name}</b></h4>
                                        <p className="secondary-txt" style={authedUser === userId ? {color: "#5885DD"}: {}}>{users[userId].id}</p>
                                    </div>
                                </div>
                            </td>
                            <td>{users[userId].questions.length}</td>
                            <td>{Object.keys(users[userId].answers).length}</td>
                        </tr>
                    ))}
                </tbody>
            </table>
        </div>
    )
}

const mapStateToProps = ({users, authedUser}) => {
    return {
        userIds: Object.keys(users).sort((a, b) => (Object.keys(users[b].answers).length + users[b].questions.length) - (Object.keys(users[a].answers).length + users[a].questions.length)),
        users,
        authedUser
    }
}
export default connect(mapStateToProps)(Leaderboard);