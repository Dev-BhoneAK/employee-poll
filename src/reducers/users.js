import {LOGIN_DATA, RECEIVE_DATA} from '../actions/shared';
import {ADD_ANSWER, ADD_QUESTION_ID, LOGOUT, SIGNUP} from '../actions/users';
const users = (state = {}, action) => {
    switch (action.type) {
        case RECEIVE_DATA:
            return{
                ...state,
                ...action.users,
            }
        case ADD_QUESTION_ID:
            return {
                ...state,
                [action.author]: {
                    ...state[action.author],
                    questions: state[action.author].questions.concat(action.id)
                }
            }
        case ADD_ANSWER:
            return {
                ...state,
                [action.authedUser]: {
                    ...state[action.authedUser],
                    answers: {
                        ...state[action.authedUser].answers,
                        [action.qid]: action.answer
                    }
                }
            }
        case SIGNUP:
            return {
                ...state,
                [action.newSignupUser.id]: action.newSignupUser
            }
        case LOGOUT:
            return null;
        default:
            return state;
    }
}

export default users;