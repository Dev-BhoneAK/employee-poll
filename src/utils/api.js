import {
    _getUsers,
    _getQuestions,
    _saveQuestion,
    _saveQuestionAnswer ,
    _signupUser
} from './_DATA.js'

export function getInitialData () {
    return Promise.all([
        _getUsers(),
        _getQuestions(),
    ]).then(([users, questions]) => {

        const newSignupUser = localStorage.getItem('newSignupUser');
        if(newSignupUser){
            const signupUser = JSON.parse(newSignupUser);
            return {
                users: { ...users, [signupUser?.id] : signupUser},
                questions,
            };
        }else{
            return {
                users,
                questions,
            }
        }
    })
}

export function saveQuestion (info) {
    return _saveQuestion(info)
}

export function saveQuestionAnswer (info) {
    return _saveQuestionAnswer(info)
}

export function getApiUsers() {
    return _getUsers();
}

export function signupUser(user){
    return _signupUser(user);
}