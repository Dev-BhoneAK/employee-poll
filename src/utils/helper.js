export function formatDate (timestamp) {
    const d = new Date(timestamp)
    const time = d.toLocaleTimeString('en-US')
    return time.substr(0, 5) + time.slice(-2) + ' | ' + d.toLocaleDateString()
}

export function formatQuestion (question, user) {
    const { id, timestamp } = question;
    const { name, avatarURL } = user

    return {
        name,
        id,
        timestamp,
        avatar: avatarURL
    }
}