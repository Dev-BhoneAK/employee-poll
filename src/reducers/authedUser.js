import {LOGIN, LOGOUT} from '../actions/users';
import {RECEIVE_DATA} from '../actions/shared';

const authedUser = (state = null, action) => {
    switch (action.type) {
        case LOGIN:
            return action.authedUser;
        case RECEIVE_DATA:
            return action.authedUser;
        case LOGOUT:
            return null;
        default:
            return state;
    }
}

export default authedUser;