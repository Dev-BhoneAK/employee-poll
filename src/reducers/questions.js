import {LOGIN_DATA, RECEIVE_DATA} from '../actions/shared';
import {ADD_QUESTION, ADD_USER_ID} from  '../actions/questions';
import {LOGOUT} from "../actions/users";

const questions = (state = {}, action) => {
    switch (action.type) {
        case LOGIN_DATA:
            return {
                ...state,
                ...action.questions
            }
        case RECEIVE_DATA:
            return {
                ...state,
                ...action.questions
            }
        case ADD_QUESTION:
            return {
                ...state,
                [action.question.id]: action.question
            }
        case ADD_USER_ID:
            return {
                ...state,
                [action.qid]: {
                    ...state[action.qid],
                    [action.answer]: {
                        ...state[action.qid][action.answer],
                        votes: state[action.qid][action.answer].votes.concat(action.authedUser)
                    }
                }
            }
        case LOGOUT:
            return null;
        default:
            return state;
    }
}

export default questions;