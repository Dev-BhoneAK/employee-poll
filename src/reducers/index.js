import {combineReducers} from "redux";
import questions from './questions';
import users from "./users";
import authedUser from "./authedUser";
import {loadingBarReducer} from "react-redux-loading-bar";

export default combineReducers({questions, users, authedUser, loadingBar: loadingBarReducer});