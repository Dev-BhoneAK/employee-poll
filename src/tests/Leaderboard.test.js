import { render, screen } from '@testing-library/react';
import Leaderboard from '../pages/Leaderboard';
import {Provider} from "react-redux";
import {createStore} from "redux";
import {MemoryRouter} from "react-router-dom";
import reducers from "../reducers";
import {_getUsers} from "../utils/_DATA";

describe('Leaderboard', () => {
    it("will match with snapshot", async () => {

        const users = await _getUsers();
        const initialState = {
            users,
            authedUser: 'mtsamis'
        }
        const store = createStore(reducers, initialState);
        // eslint-disable-next-line testing-library/render-result-naming-convention
        const leaderboard = render(
            <Provider store={store}>
                <MemoryRouter>
                    <Leaderboard />
                </MemoryRouter>
            </Provider>
        );
        expect(leaderboard).toMatchSnapshot();
    })
})