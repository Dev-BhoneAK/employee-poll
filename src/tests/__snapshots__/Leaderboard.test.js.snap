// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard will match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="table-container"
      >
        <table>
          <thead>
            <tr>
              <th>
                Users
              </th>
              <th>
                Created
              </th>
              <th>
                Answered
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <div
                  class="user-data"
                >
                  <img
                    alt="Avatar"
                    class="leaderboard-avatar"
                    src="/images/employee2.jpg"
                  />
                  <div>
                    <h4>
                      <b>
                        Sarah Edo
                      </b>
                    </h4>
                    <p
                      class="secondary-txt"
                    >
                      sarahedo
                    </p>
                  </div>
                </div>
              </td>
              <td>
                2
              </td>
              <td>
                4
              </td>
            </tr>
            <tr
              style="color: rgb(88, 133, 221);"
            >
              <td>
                <div
                  class="user-data"
                >
                  <img
                    alt="Avatar"
                    class="leaderboard-avatar"
                    src="/images/employee1.jpg"
                  />
                  <div>
                    <h4>
                      <b>
                        Mike Tsamis
                      </b>
                    </h4>
                    <p
                      class="secondary-txt"
                      style="color: rgb(88, 133, 221);"
                    >
                      mtsamis
                    </p>
                  </div>
                </div>
              </td>
              <td>
                2
              </td>
              <td>
                3
              </td>
            </tr>
            <tr>
              <td>
                <div
                  class="user-data"
                >
                  <img
                    alt="Avatar"
                    class="leaderboard-avatar"
                    src="/images/employee3.jpg"
                  />
                  <div>
                    <h4>
                      <b>
                        Tyler McGinnis
                      </b>
                    </h4>
                    <p
                      class="secondary-txt"
                    >
                      tylermcginnis
                    </p>
                  </div>
                </div>
              </td>
              <td>
                2
              </td>
              <td>
                2
              </td>
            </tr>
            <tr>
              <td>
                <div
                  class="user-data"
                >
                  <img
                    alt="Avatar"
                    class="leaderboard-avatar"
                    src="/images/employee4.jpg"
                  />
                  <div>
                    <h4>
                      <b>
                        Zenobia Oshikanlu
                      </b>
                    </h4>
                    <p
                      class="secondary-txt"
                    >
                      zoshikanlu
                    </p>
                  </div>
                </div>
              </td>
              <td>
                0
              </td>
              <td>
                1
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="table-container"
    >
      <table>
        <thead>
          <tr>
            <th>
              Users
            </th>
            <th>
              Created
            </th>
            <th>
              Answered
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <div
                class="user-data"
              >
                <img
                  alt="Avatar"
                  class="leaderboard-avatar"
                  src="/images/employee2.jpg"
                />
                <div>
                  <h4>
                    <b>
                      Sarah Edo
                    </b>
                  </h4>
                  <p
                    class="secondary-txt"
                  >
                    sarahedo
                  </p>
                </div>
              </div>
            </td>
            <td>
              2
            </td>
            <td>
              4
            </td>
          </tr>
          <tr
            style="color: rgb(88, 133, 221);"
          >
            <td>
              <div
                class="user-data"
              >
                <img
                  alt="Avatar"
                  class="leaderboard-avatar"
                  src="/images/employee1.jpg"
                />
                <div>
                  <h4>
                    <b>
                      Mike Tsamis
                    </b>
                  </h4>
                  <p
                    class="secondary-txt"
                    style="color: rgb(88, 133, 221);"
                  >
                    mtsamis
                  </p>
                </div>
              </div>
            </td>
            <td>
              2
            </td>
            <td>
              3
            </td>
          </tr>
          <tr>
            <td>
              <div
                class="user-data"
              >
                <img
                  alt="Avatar"
                  class="leaderboard-avatar"
                  src="/images/employee3.jpg"
                />
                <div>
                  <h4>
                    <b>
                      Tyler McGinnis
                    </b>
                  </h4>
                  <p
                    class="secondary-txt"
                  >
                    tylermcginnis
                  </p>
                </div>
              </div>
            </td>
            <td>
              2
            </td>
            <td>
              2
            </td>
          </tr>
          <tr>
            <td>
              <div
                class="user-data"
              >
                <img
                  alt="Avatar"
                  class="leaderboard-avatar"
                  src="/images/employee4.jpg"
                />
                <div>
                  <h4>
                    <b>
                      Zenobia Oshikanlu
                    </b>
                  </h4>
                  <p
                    class="secondary-txt"
                  >
                    zoshikanlu
                  </p>
                </div>
              </div>
            </td>
            <td>
              0
            </td>
            <td>
              1
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
